pynitrokey/cli/nk3/test.py

```diff
@@ -0,0 +1,189 @@
+
+    create_options, state = server.register_begin(
+        user, user_verification=uv, authenticator_attachment="cross-platform"
+    )
+
+    local_print("Please press the touch button on the device ...")
+    make_credential_result = client.make_credential(create_options["publicKey"])
+    cert = make_credential_result.attestation_object.att_statement["x5c"]
+    cert_hash = sha256(cert[0]).digest().hex()
+    data = ""
+    if cert_hash != FIDO2_CERT_HASH:
+        data = cert_hash
+
+    auth_data = server.register_complete(
+        state,
+        make_credential_result.client_data,
+        make_credential_result.attestation_object,
+    )
+    credentials = [auth_data.credential_data]
+
+    request_options, state = server.authenticate_begin(
+        credentials, user_verification=uv
+    )
+
+    local_print("Please press the touch button on the device ...")
+    get_assertion_result = client.get_assertion(request_options["publicKey"])
+    get_assertion_response = get_assertion_result.get_response(0)
+
+    server.authenticate_complete(
+        state,
+        credentials,
+        get_assertion_response.credential_id,
+        get_assertion_response.client_data,
+        get_assertion_response.authenticator_data,
+        get_assertion_response.signature,
+    )
+
+    return TestResult(TestStatus.SUCCESS, data)
+
+
+def run_tests(device: Nitrokey3Base) -> bool:
+    results = []
+
+    local_print("")
+    local_print(f"Running tests for {device.name} at {device.path}")
+    local_print("")
+
+    n = len(TEST_CASES)
+    idx_len = len(str(n))
+    name_len = max([len(test_case.name) for test_case in TEST_CASES]) + 2
+    status_len = max([len(status.name) for status in TestStatus]) + 2
+
+    for (i, test_case) in enumerate(TEST_CASES):
+        try:
+            result = test_case.fn(device)
+        except Exception:
+            result = TestResult(TestStatus.FAILURE, exc_info=sys.exc_info())
+        results.append(result)
+
+        idx = str(i + 1).rjust(idx_len)
+        name = test_case.name.ljust(name_len)
+        status = result.status.name.ljust(status_len)
+        msg = ""
+        if result.data:
+            msg = str(result.data)
+        elif result.exc_info:
+            logger.error(
+                f"An exception occured during the execution of the test {test_case.name}:",
+                exc_info=result.exc_info,
+            )
+            msg = str(result.exc_info[1])
+
+        local_print(f"[{idx}/{n}]\t{name}\t{status}\t{msg}")
+
+    success = len([result for result in results if result.status == TestStatus.SUCCESS])
+    skipped = len([result for result in results if result.status == TestStatus.SKIPPED])
+    failed = len([result for result in results if result.status == TestStatus.FAILURE])
+    local_print("")
+    local_print(f"{n} tests, {success} successful, {skipped} skipped, {failed} failed")
+
+    return not any([result.status == TestStatus.FAILURE for result in results])
```

Comment:
nit: for readability and cognition load I think it would be better to check for `all()` instead of  `not any()`, "positive" logic so to speak

Reply:
sure, I will update it

Reply:
done